src/System.Private.ServiceModel/src/System/IdentityModel/SignatureResourcePool.cs

```diff
@@ -2,17 +2,56 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-
 using System.IO;
 using System.Security.Cryptography;
 using System.IdentityModel.Tokens;
 using System.Xml;
 using System.Text;
+using System.ServiceModel;
 
 namespace System.IdentityModel
 {
     // for sequential use by one thread
     internal sealed class SignatureResourcePool
     {
+        private HashStream _hashStream;
+        private HashAlgorithm _hashAlgorithm;
+
+        private HashAlgorithm TakeHashAlgorithm(string algorithm)
+        {
+            if (_hashAlgorithm == null)
+            {
+                if (string.IsNullOrEmpty(algorithm))
+                {
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgument(algorithm, SR.Format(SR.EmptyOrNullArgumentString, nameof(algorithm)));
+                }
+
+                _hashAlgorithm = CryptoHelper.CreateHashAlgorithm(algorithm);
+            }
+            else
+            {
+                _hashAlgorithm.Initialize();
+            }
+
+            return _hashAlgorithm;
```

Comment:
This doesn't look like it's "taking" so much as "borrowing" or "sharing".

Reply:
I agree. Copied code from .NET Framework and that was the name there. I wanted to maintain the code the same as much as possible to reduce mental gymnastics for servicing.

---
In reply to: [274010183](https://github.com/dotnet/wcf/pull/3538#discussion_r274010183) [](ancestors = 274010183)